### ISSUE TITLE
snap_usedef can eliminate GC anchors

### ISSUE THREAD

**Issue by corsix** [LuaJIT/LuaJIT#1167]
Inspired by [a HN thread](https://news.ycombinator.com/item?id=39509927), there is an expectation that local variables act as GC anchors for FFI cdata. Unfortunately, `snap_usedef` can shorten the lifetime of local variables, thereby eliminating local variables that were serving as GC anchors, as in:

```lua
local ffi = require"ffi"
local function collect_lots()
  for i = 1, 20 do collectgarbage() end
end
local function f(s)
  local blob = ffi.new"int[2]"
  local interior = blob + 1
  interior[0] = 13 -- should become the return value
  s:gsub(".", collect_lots)
  return interior[0] -- kept alive by blob?
end
for i = 1, 60 do
  local str = ("x"):rep(i - 59)
  local n = f(str)
  assert(n == 13, n) -- can fail!!
end
```

As a less convincing example, a behavioural difference between JIT-on and JIT-off can be observed by using the debug library to inspect a stack slot which is otherwise dead:
```lua
local function inspect_stack()
  local name, val = debug.getlocal(3, 2)
  assert(name == "secret")
  assert(val == "x") -- can fail
end
local function f(s)
  local secret = s
  local interior = {}
  interior[0] = 13
  s:gsub(".", inspect_stack)
end
for i = 1, 60 do
  local str = ("x"):rep(i - 59)
  f(str)
end
```

**Comment by trans07** [comment]
> Inspired by [a HN thread](https://news.ycombinator.com/item?id=39509927), there is an expectation that local variables act as GC anchors for FFI cdata.

https://luajit.org/ext_ffi_semantics.html
```
ffi.cdef[[
typedef struct { int *a; } foo_t;
]]

local s = ffi.new("foo_t", ffi.new("int[10]")) -- WRONG!

local a = ffi.new("int[10]") -- OK
local s = ffi.new("foo_t", a)
-- Now do something with 's', but keep 'a' alive until you're done.
```
What is "keeping 'a' alive" here if not the `local a =` ?

**Comment by MikePall** [comment]
The fundamental problem here is that developer expectations do not meet VM reality.

From the perspective of the GC, an object is kept 'alive'  by its *use*. And not by a variable declaration or the extent of its scope. Also, the JIT compiler has a much narrower definition of what is considered a use.

That said, this is a non-issue with plain Lua data structures. The language does not allow interior pointers. And reference semantics prevent aliasing issues.

So we're mainly talking about the FFI. But similar issues pop up with internal optimizations, e.g. on string buffers (look for `IR_USE`).

Now, what can we do:

First, the documentation needs to be reworded. It should avoid talking about stacks, because this mental model does not fully match how the VM works right now. Much less so in a (potential) future VM using SSA bytecode.

Second, how can we improve the developer experience?

It would be relatively easy to offer a first-class API like `ffi.keep(obj)`, which compiles down to `IR_USE`. This would come with semantic guarantees, so future improvements to the JIT compiler will not foil the intent.

The problem is that `ffi.keep(obj)` needs to be *at the end* of the desired minimum lifetime. The name could lead to misunderstandings, too. OTOH `ffi.use(obj)` wouldn't be much better to make its correct use clear.

It's also uncomfortable to use, because it will often look like this:
```Lua
local p = clib.get(obj)
local tmp = ffi.string(p)
ffi.keep(obj)
return tmp
```

This could be ameliorated by passing through all arguments except the first one:
```Lua
local p = clib.get(obj)
return ffi.keep(obj, ffi.string(p))
```
But that looks weird. And it's not self-explanatory how and why this works, either.

Another thing to consider is that even innocent-looking code is not completely safe:
```Lua
local c2 = ffi.new("char[2]")
return ffi.string(c2 + 1, 1)
```
Which is really:
```Lua
local c2 = ffi.new("char[2]")
local tmp = c2 + 1
return ffi.string(tmp, 1)
```
The GC could potentially collect `c2` right before the call to `ffi.string()`.

The natural mechanism to keep an object alive would be syntactic scope. Only function scopes are feasible without substantial changes:
```Lua
return ffi.with(obj, function(obj)
  local p = clib.get(obj)
  return ffi.string(p)
end)
```
Apart from efficiency considerations and NYI issues due to the closure allocation, I still don't like it.

And to get back to the beginning: I guess the only way to match developer expectations is a kind of 'keep' attribute that extends from the declaration to the end of its scope. A 'magic' function that pins its resulting stack slot? Umm, no. I don't see a clean way to implement this without doing it at the syntactic level.

**Comment by trans07** [comment]
> The fundamental problem here is that developer expectations do not meet VM reality.

Forget about the admittedly wonky reference to stacks, I'm still left trying to understand what the appropriate expectation about cdata lifetimes was to be gleaned from that verbatim code snippet. 

> Also, the JIT compiler has a much narrower definition of what is considered a use.

It would probably be useful to document what those are? Does stuffing the cdata in a table value suffice? 

I imagine it will be a long time coming before an FFI call isn't a guaranteed pin. At this time it is safe to say that: `void keep(void *) {}` is the best workaround?

**Comment by CapsAdmin** [comment]
> > The fundamental problem here is that developer expectations do not meet VM reality.
> 
> Forget about the admittedly wonky reference to stacks, I'm still left trying to understand what the appropriate expectation about cdata lifetimes was to be gleaned from that verbatim code snippet.
> 
> > Also, the JIT compiler has a much narrower definition of what is considered a use.
> 
> It would probably be useful to document what those are? Does stuffing the cdata in a table value suffice?
> 
> I imagine it will be a long time coming before an FFI call isn't a guaranteed pin. At this time it is safe to say that: `void keep(void *) {}` is the best workaround?

As far as I understand in the first example just by reading the code, (_though it took a while for me to understand what's actually going on_)

 `blob` is the actual thing allocated and `interior` is just a view to `blob` offset right by 1. The fact that it's offset right by 1 is not relevant, it could be offset by 0 or just `ffi.cast("int[2]", blob)`

Right after `interior` is created, `blob` is no longer **used**/needed and can be collected at any time, so the `interior` view to `blob` cannot be guaranteed anymore.

This is covered in the docs under "Garbage Collection of cdata Objects"

So the question now is, when should it be collected? At the end of scope or as soon as it's no longer used?

Ignoring technical challenges, in this example I think it's easy to see that it should be at the end of scope, but that may be a bad tradeoff in other cases. Ideally cdata should just depend on other cdata it was created from, but I guess that has performance and/or complexity problems.

**Comment by CapsAdmin** [comment]
> The problem is that `ffi.keep(obj)` needs to be _at the end_ of the desired minimum lifetime. The name could lead to misunderstandings, too. OTOH `ffi.use(obj)` wouldn't be much better to make its correct use clear.

My first reaction to seeing keep is that I would think it's kept until it's unkept somehow. Some simple manual control like ffi.pin and ffi.unpin would feel natural to me given how the ffi library is "no hand holding" and low level.

**Comment by trans07** [comment]
> This is covered in the docs under "Garbage Collection of cdata Objects"
> 
> So the question now is, when should it be collected? Is this surprising behavior?

It is surprising because of that same section in the manual:

```lua
local s = ffi.new("foo_t", ffi.new("int[10]")) -- WRONG!
```
then this:
```lua
local a = ffi.new("int[10]") -- OK
```

What is the difference between those two lines, and is why is the second one `OK`?

There is the further comment: "Now do something with 's', but keep 'a' alive until you're done."
But with nothing further defining what keep 'a' alive means what is one left to believe? Without further definition it is reasonable that the preceding difference between WRONG and OK is what is essential. Also I never experienced any deviation in this behavior from 2.0.
The provided documentation is what makes this surprising, not some mistaken intuition.
Other FFI systems have keep primitives so its omission is also suggestive.

In any event, `ffi.keep` isn't an interface that people in this space will be unfamiliar with (for ex .NET). Eventually one accepts that FFI and elegance do not belong in the same sentence. The thing is, in the situations where keep would be necessary you are only going to guarantee correctness with an equally ugly hack (like an empty keep function).

I also agree any expensive solution is inappropriate as not every FFI use case encounters this problem.

> > It's up to you somehow to store a somewhere. If this is a function that returns some table object, you could store it in the table in a private field.

Does that solve anything? That just kicks the can down the road. If I create a table simply for storing references then don't I have to "use" that table value if not the table. If the VM in the future observes the table value is never referred to, then it is just as fragile.
These things need to be defined.
With no prior assumptions a['k'] = b and local a = b are not any different.

**Comment by MikePall** [comment]
A 'pin' method already has an established and different meaning. It's used to mark an object as unmovable, i.e. its memory address stays pinned. OK, so LuaJIT has a non-moving GC, so this is a not an issue right now. And cdata objects can probably never be made movable, since this would necessitate tons of pin annotations everywhere.

Also, the start of a lifetime doesn't really need to be spelled out. The variable or parameter declaration is always the start of the lifetime.

What we need is a way to mark the *end of the lifetime* in a convenient and intuitive manner. The best place for this might be the point of declaration, though.

**Comment by MikePall** [comment]
The 'dependency' solution would be to allow GC references from an interior pointer to a cdata object. This is another can of worms, which I didn't want to open above.

Implicit GC references would necessitate all cdata objects to have some extra space for these references. And for the VM to set these references. And for the GC to check and traverse them. This has a negative performance impact for all FFI objects. Also, one would still need to declare or set an explicit dependency for the typical case of C accessor functions.

Explicit GC references can be implemented (by hand even right now) with a global weak table. But this table will grow fast, when explicit references are used in a critical path. And weak tables are really expensive. They need to be traversed multiple times by the GC and in the atomic phase, too. The latter is bad for maximum GC latency.

I'm still leaning towards a scope-based solution, since this probably leads to a zero-overhead implementation.

**Comment by CapsAdmin** [comment]
> What is the difference between those two lines, and is why is the second one `OK`?

In the wrong example there is no way to store `a`, but in the OK example it's implied that you should imagine how to store `a`. I think that was the intended purpose of the example. I think it could benefit from some more involved examples.

Check out this gist which has a similar example but in a more complete way with more wrong and correct use cases

https://gist.github.com/CapsAdmin/2edde2741e7715a9d64c597ed2d9b139

I feel once you understand this, it's not very surprising anymore (unless I'm missing something from your point of view). But I'll agree that getting to understand this in the first place can be confusing.

**Comment by trans07** [comment]
> In the wrong example there is no way to store `a`, but in the OK example it's implied that you should imagine how to store `a`.

No. What does "storing `a`" mean? That would need a clear definition. This is about lifetime and what are references, not "storage".

Your examples are wrong and proves my point. Line 82 is precisely what is under discussion. The whole issue is that line 83 is not enough to keep that array from being GC'ed. `a` is clearly never used past 83, so the VM is in the clear to collect it.

Your example marked OK starting from line 44 is also not correct: `local a, s = get_s()`, and then the variable `a` is never used again, so the GC is free to collect it, there is no guarantee after line 51.
Side note those examples won't be JITed because of the use of closures. So the fact that they seem to work isn't really proving anything related to this issue.

The example from line 63, while it works, isn't fundamentally different. It so happens that LuaJIT isn't clever enough to notice that `__a` is never used, but there is no reason it couldn't, since you can plainly prove that table entry is never used.

**Comment by CapsAdmin** [comment]
@trans07 The last example was not supposed to be there, my bad. But you're absolutely right about example at line 44. It's also an interesting observation that this can potentially be causing issues in the future if LuaJIT decides to be clever about tracking fields of a table.

The documentation right now states that you should keep the upvalue alive somehow (I just decided to use the term store) so in a way this hinges on it not being optimized away somehow (or in the future). I guess the only sane way I can see from your perspective as I understand it is to ultimately store it globally somewhere.

**Comment by trans07** [comment]
> The documentation right now states that you should keep the upvalue alive somehow 

I think we've established that in terms of LuaJIT 2.1, the documentation is simply incorrect.
I don't think upvalues are a first class concept in the VM. The JIT would be free to elide them and then you're back at square one. The only reason this doesn't happen to be a problem today is because closures are NYI.

> I guess the only sane way I can see from your perspective as I understand it is to ultimately store it globally somewhere.

Absolutely not, even a global store might not be sufficient if you don't use it again and it is usually a performance disaster as noted. In general the simple lifetime until end of use, full stop, is reasonable in a garbage collected language like Lua but then this problem comes up and is nothing unique to LuaJIT.
Yes technically, `keep` or similar is storing a reference in a set of continuations, which I think is the best place for it regardless of the syntactic affordance to declare it, but this is sort of unconventional way of describing it. In any case, store in the context of use would need to be defined, and my take from Mike Pall's posts are that generally a binding or a data structure is not it.

**Comment by corsix** [comment]
> Does stuffing the cdata in a table value suffice?

You've then got to keep the table alive somehow. As an example of subtle points, the following exhibits the same problem as the original:
```lua
local ffi = require"ffi"
local function collect_lots()
  for i = 1, 20 do collectgarbage() end
end
local function f(s)
  local blob = ffi.new"int[2]"
  local interior = blob + 1
  blob = {anchor = blob} -- stuffing it in a table
  interior[0] = 13 -- should become the return value
  s:gsub(".", collect_lots)
  return interior[0] -- kept alive by blob?
end
for i = 1, 60 do
  local str = ("x"):rep(i - 59)
  local n = f(str)
  assert(n == 13, n) -- can fail!!
end
```

> I imagine it will be a long time coming before an FFI call isn't a guaranteed pin. At this time it is safe to say that: `void keep(void *) {}` is the best workaround?

That would do it, though something cheaper that comes to mind is `do local _ = blob end`, as in:
```lua
local ffi = require"ffi"
local function collect_lots()
  for i = 1, 20 do collectgarbage() end
end
local function f(s)
  local blob = ffi.new"int[2]"
  local interior = blob + 1
  interior[0] = 13 -- should become the return value
  s:gsub(".", collect_lots)
  local result = interior[0]
  do local _ = blob end -- keeps blob alive
  return result
end
for i = 1, 60 do
  local str = ("x"):rep(i - 59)
  local n = f(str)
  assert(n == 13, n)
end
```

> The 'dependency' solution would be to allow GC references from an interior pointer to a cdata object. This is another can of worms, which I didn't want to open above.

There are production GCs, like Go's, which allow interior pointers to keep the enclosing allocation alive. This is only helpful for pointers strictly inside the bounds of the enclosing allocation though, and [the original motivating example](https://wingolog.org/archives/2024/02/26/on-the-impossibility-of-composing-finalizers-and-ffi) was something like a `hb_blob_get_data` call, the result of which might be "owned" by the cdata, but be a completely different allocation. All that is to say that I'm also not convinced by a 'dependency' solution.

**Comment by trans07** [comment]
> You've then got to keep the table alive somehow. As an example of subtle points, the following exhibits the same problem as the original:

Well yes, the was more rhetorical to illustrate the ambiguity with the docs.

> That would do it, though something cheaper that comes to mind is `do local _ = blob end`, as in:
> 
>   do local _ = blob end -- keeps blob alive

Temporary workaround maybe, but an unused assignment (a total NOP) is just as fragile, no? I'm a little surprised that even works today. The difference with `ffi.C.keep` was as noted it will be a long time coming before the invariant that a C FFI call ensures its arguments are alive doesn't hold. A C function call is better than maintaining a weak table of references.

**Comment by sonoro1234** [comment]
This is the pattern I use to apply for ffi gc-anchors.
It does not assert but after reading the whole thread I am compeled to ask if this is a correct aproach.

```Lua
local ffi = require"ffi"
local function collect_lots()
  for i = 1, 20 do collectgarbage() end
end
local anchors = {}
local function f(s)
  local blob = ffi.new"int[2]"
  anchors[tostring(blob)] = blob
  local interior = blob + 1
  interior[0] = 13 -- should become the return value
  s:gsub(".", collect_lots)
  return interior[0] -- kept alive by blob?
end
for i = 1, 60 do
  local str = ("x"):rep(i - 59)
  local n = f(str)
  assert(n == 13, n) -- can fail!!
end
--release and test anchors
for k,v in pairs(anchors) do
	assert(k == tostring(v))
end
anchors = nil
```

**Comment by jakitliang** [comment]
> This is the pattern I use for ffi gc-anchors. It does not assert but after reading the whole thread I am compeled to ask if this is a correct aproach.
> 
> local ffi = require"ffi"
> local function collect_lots()
>   for i = 1, 20 do collectgarbage() end
> end
> local anchors = {}
> local function f(s)
>   local blob = ffi.new"int[2]"
>   anchors[tostring(blob)] = blob
>   local interior = blob + 1
>   interior[0] = 13 -- should become the return value
>   s:gsub(".", collect_lots)
>   return interior[0] -- kept alive by blob?
> end
> for i = 1, 60 do
>   local str = ("x"):rep(i - 59)
>   local n = f(str)
>   assert(n == 13, n) -- can fail!!
> end
> --release and test anchors
> for k,v in pairs(anchors) do
> 	assert(k == tostring(v))
> end
> anchors = nil
> If the code below `--release and test anchors` is commented, the `scope condition` is still valid for the GC but not the `use condition`.
> 
> I should say also that the assert in the first code by [@corsix](https://github.com/corsix) only happens to me with 64bit LuaJIT but not with 32bit.

Release build of LuaJIT will ignore the `assert` so it won't trigger the MSVC assert error messagebox.

**Comment by Dr-Emann** [comment]
Go uses the name [`runtime.KeepAlive`](https://pkg.go.dev/runtime#KeepAlive), I think the name `ffi.keepalive` seems to do a reasonable job to me of explaining what it does.

**Comment by rweichler** [comment]
My tl;dr of this thread: there are a couple workarounds but they are dependent on LuaJIT's current implementation and could break in the future. It needs to be an official LuaJIT feature.

trans07 method
```lua
gcanchor = ffi.cast('void (*)(void *)', function() end)

local p = clib.get(obj)
local tmp = ffi.string(p)
gcanchor(obj)
return tmp
```

corsix method
```lua
local p = clib.get(obj)
local tmp = ffi.string(p)
do local gcanchor = obj end 
return tmp
```

My approach is to just pick a pattern/name and stick with it, and when the time comes that this becomes an official feature I'll do a global search + replace on the string in my codebase.

**Comment by Desour** [comment]
I'd like to note that oftentimes one does not only need a value to live at least as long as one works with it, but also that eventually the values will be destructed in a certain order (i.e. due to destructors accessing other objects).

Example:
```lua
local ffi = require("ffi")

ffi.cdef([[
struct Obj { struct Obj **me_in_parent; };
struct Ctx { struct Obj *obj[1]; };
/* (Ctx::obj is just array to allow taking its address) */
]])

do
	local ctx = ffi.new("struct Ctx")

	local obj_a = ffi.gc(ffi.new("struct Obj"), function(o)
		o.me_in_parent[0] = nil -- remove myself from parent ctx
	end)
	ctx.obj[0] = obj_a
	obj_a.me_in_parent = ctx.obj
end

-- ctx and obj_a can be destructed in any order.
-- if obj_a is destructed first, it leads to a UaF
```

(One solution would be to wrap the Ctx struct in a refcounted ptr that you alloc with malloc/free.)

One could use an anchor in the finalizer though:
```lua
	local obj_a = ffi.gc(ffi.new("struct Obj"), function(o)
		o.me_in_parent[0] = nil -- remove myself from parent ctx
		ffi.gcanchor(ctx)
	end)
```
This means `obj_a`'s finalizer would hold `ctx` as upvalue.

As creating closures is NYI, one might want to replace it with a table with a metatable:
```lua

local mt = {__call = function(self, o)
		o.me_in_parent[0] = nil
		ffi.gcanchor(self[0])
end}
local struct_Obj_finalizer_for_ctx = setmetatable({ctx}, mt) -- (can share between Objs that use the same Ctx)
local obj_a = ffi.gc(ffi.new("struct Obj"), struct_Obj_finalizer_for_ctx)
```

But now, `ctx` is captured in the finalizer anyways, so the `ffi.gcanchor` becomes irrelevant.
(Maybe this was already the case with the closure above. Idk if luajit is allowed to optimize away upvalues, considering there's a `debug` api that can access upvalues.)

TL;DR: You can build dependencies by putting references into finalizers.

Idk if this is efficient though, especially if you build long chains. Maybe it's even worse than weak tables?
